proposals/p6333.md

```diff
@@ -0,0 +1,586 @@
+        -   `a.carbon` and `b.carbon` will be available for imports, but it
+            should be an error if `b.carbon` is imported directly. This is
+            required because `a.carbon` can expose `b.carbon` on the import
+            boundary, meaning an indirect import of `b.carbon` must work..
+    -   Link object files into an executable.
+
+For both, there should be an implicit dependency on the full Core package, not
+just the prelude. This is because we want the Core package to be easy to access.
+
+#### Indirect API exposure
+
+The `apis` attribute is suggested to support only _direct_ dependencies. For
+example:
+
+```bazel
+carbon_library(
+    name = "a",
+    apis = ["a.carbon"],
+)
+carbon_library(
+    name = "b",
+    apis = ["b.carbon"],
+    deps = [":a"],
+)
+carbon_library(
+    name = "c",
+    srcs = ["c.carbon"],
+    deps = [":b"],
+)
+```
+
+If `c.carbon` imports `a.carbon`, the build should error that `a.carbon`
+requires a direct dependency. We should allow forwarding, so that the same could
+compile without requiring `c` to have a direct dependency on `a`. This should
+look like `public_deps = [":a"]`, added to `b` (and superseding the need to list
+`:a` in `deps`).
+
+This feature may see frequent use, for example in `Core` to allow writing it as
+multiple libraries instead of one large glob. But it's probably also something
+that can be delayed a little, because we can just use a big glob and force
+direct dependencies.
+
+#### Core package rules
+
+In the `core/` directory, we will set up corresponding `carbon_library` rules.
+These will need to pass flags to opt-out of normal behaviors, in particular the
+dependency on the prelude library.
+
+### Mapping packaging directives to filenames
+
+Suppose we have some command line `carbon compile a.carbon`, and in `a.carbon`,
+it does `import Core library "map";`. This needs to load `core/map.carbon`, and
+without parsing every file matching `core/**/*.carbon`.
+
+In order to achieve this:
+
+-   The `compile` command will have a built-in directory mapping for the `Core`
+    package, for example to `/usr/lib/carbon/core` (when installed to the `/usr`
+    prefix).
+-   The `map` library name will need to match the filename, so
+    `/usr/lib/carbon/core/map.carbon`.
+    -   Slashes may be provided in the library name, for subdirectories.
+-   If `map.carbon` has other `Core` imports, they will be recursively loaded
+    once parsed.
+    -   Note that checking isn't required to process imports from a file.
+-   When determining which `impl` files need to be associated with a given
+    `api`, we will support both `<library name>.impl.carbon` and
+    `<library name>.*.impl.carbon`, with the latter supporting arbitrary names
+    for cases where there are multiple impl files.
+
+Because we'll build this for Core, it would probably be straightforward to
+expose this for other packages, too. So for example, we will support
+`--package_path=MyPackage:/my/package`.
```

Comment:
Do we want some way to specify whether a package is precompiled or not (that is, whether `carbon build` should go looking for `impl` files and build and link them, or whether we should instead link against a corresponding `.a` file found somewhere)? Having our behavior depend on whether the `impl` files happen to exist on disk sounds a little fragile for an environment like `bazel` that may or may not run the build in a cut-down tree with fewer files present -- but that would be a non-issue if `bazel` is using `carbon compile` instead of `carbon build`, since the former presumably only goes looking for api files, not for impl files.